### PR TITLE
Use AsyncWriterPool for Fastq writing.

### PR DESCRIPTION
**Pull request by jacarey**
### Description

This PR adds the new AsyncWriterPool from htsjdk 2.24.0 for fastq writing. 

Additionally, it removes some of the write threading from the basecalls converters since this is now covered in the async writing and would cause thread thrashing if left as it was.

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [x] Added or modified tests to cover changes and any new functionality
- [x] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

File: src/main/java/picard/illumina/IlluminaBasecallsToFastq.java [hunk 1]
```diff
@@ -196,26 +199,40 @@
...
 
     @Override
     protected int doWork() {
         initialize();
         final Set<String> barcodes = sampleBarcodeClusterWriterMap.keySet();
-        basecallsConverter.processTilesAndWritePerSampleOutputs(barcodes);
+        try {
+            basecallsConverter.processTilesAndWritePerSampleOutputs(barcodes);
+        } catch (IOException e) {
+            return 1;
```
Comment (gbggrant) [inline src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:219]: Do you want to rethrow a PicardException here with some context-specific error information?

File: src/main/java/picard/illumina/IlluminaBasecallsToSam.java [hunk 2]
```diff
@@ -258,7 +259,11 @@
     @Override
     protected int doWork() {
         initialize();
-        basecallsConverter.processTilesAndWritePerSampleOutputs(barcodeSamWriterMap.keySet());
+        try {
+            basecallsConverter.processTilesAndWritePerSampleOutputs(barcodeSamWriterMap.keySet());
+        } catch (IOException e) {
+            return 1;
```
Comment (gbggrant) [inline src/main/java/picard/illumina/IlluminaBasecallsToSam.java:265]: Maybe rethrow the exception as a PicardException with more information about what might have gone wrong?

**Comment by gbggrant** [comment]
@jacarey looks good as far as I can tell - but a couple of tests are failing.  We'll wait for those to be cleaned up before proceeding.  Our main concern is that there is enough test coverage for the new changes you've put in place.

**Review by gbggrant — COMMENTED** [review]
A couple of tests still failing.  Will wait for those to pass before proceeding.  Do you need any more test coverage for this?

**Review by gbggrant — APPROVED** [review]
Looks good @jacarey if you update to master we can go ahead and merge it.

**Comment by gbggrant** [comment]
@jacarey - other than "retest this please"?

**Comment by gbggrant** [comment]
retest this please

**Comment by gbggrant** [comment]
@jacarey I logged into Travis and reran that test, which passed.  good to merge?

**Comment by jacarey** [comment]
@gbggrant :+1:

**Pull request merged**